### DISCUSSION TITLE
Constraint Axis : Significance given individual constraints

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#109] [Q&A] [answered]
I had checked on the C++ repo discussions, but did not find extensive documentation on the axis significance in all Constraint Settings. The source comments gave general ideas.

General understanding:
Most constraints involve a constraint rotation space defined via 2 vectors that are used to infer a 3rd axis, and initialize one local-3-axis space reference per body. In some of these cases, one or more axis direction is significant and does something useful, but in several cases the other axis are not special other than for math assisting locking rotation (nothing special happens in that direction). 

I know that the relative relationship between these two axis is important, ex: binding this setup will cause a constraint pressure to force the fixed objects to act relative to each other.

General question:

Is there any practical usage for using non-orthogonal local-space vector combinations? Less than 90 degrees difference per the two vectors? Could that ever do anything useful? 

If they should always be ortho vectors, I think in most cases I can even just store the two-axis vectors in my files as a Quaternion rotation of stock X/Y axis with predefined assignment, ex: quat applied to (1,0,0) & (0,1,0). Would this make sense? Ex: slider + normal, hinge + normal, twist + plane

My current understood usage of the axis on constraints:

## Fixed:
I believe the two provided axis, and thus all 3 resulting axis, are purely to setup initial reference between the two bodies, and there is no real relationship or practical usage to the particular assignment or values? Possibly you can just always give global X & Y and I don’t know if any other values would be distinguishable, or always randomly generate any other random two ortho axis?
If the axis don't match, it looks like it will force a rotation on any non-static object:

settings.mAxisX1.Set(0,0,1);
settings.mAxisY1.Set(1,0,0);

settings.mAxisX2.Set(1,0,0);
settings.mAxisY2.Set(0,0,1);

## Hinge:
 2 significant axis given. Hinge axis, and the normal axis: the zero-angle axis, a reference required for defining rotation angles and motor control angle-values

if these axis are different, and one is static, the objects will first orient rotations, then act as expected in the non-fixed axis directions. If they are both dynamic, the object may sometimes orient itself unexpectedly in mid-air. 

## Slider: normal + slider
I could not find a practical usage for normal axis. Possibly any random orthogonal axis to the slide axis would work and be indistinguishable from any other ortho vector?

if these axis are different, and one is static, the objects will first orient rotation, then the non-static will support sliding in it's respective axis. If they are both dynamic, the objects will orient in mid-air like fixed constraint. 

## Swing Twist:
These appear to be structured as orthogonal axis, via plane and twist axis, since internally they calculate a 3rd normal axis. Every provided axis serves a purpose (but still could possibly be defined via a Quat and stock X/Y vectors)

if these axis are different, and one is static, the objects will first orient rotation, then the non-static will support swing-twist in it's respective axis. If they are both dynamic, the objects will orient in mid-air like upside-down fixed constraint. 

```javascript
import initJolt from './js/jolt-physics.wasm-compat.js';

initJolt().then(function (Jolt) {
// Initialize this example
initExample(Jolt, null);

// Create a basic floor
createFloor();

function setupBoxes(offX, offY, offZ, type, axis, static1, static2) {
	const motion1 = static1 ? Jolt.EMotionType_Static : Jolt.EMotionType_Dynamic;
	const motion2 = static2 ? Jolt.EMotionType_Static : Jolt.EMotionType_Dynamic;
	const layer1 = static1 ? LAYER_NON_MOVING : LAYER_MOVING;
	const layer2 = static2 ? LAYER_NON_MOVING : LAYER_MOVING;
	const box1 = createBox(new Jolt.RVec3(-5 + offX, 7 + offY, -15+offZ), Jolt.Quat.prototype.sIdentity(), new Jolt.Vec3(3, 1, 1), motion1, layer1);
	const box2 = createBox(new Jolt.RVec3(5 + offX, 7 + offY, -15+offZ), Jolt.Quat.prototype.sIdentity(), new Jolt.Vec3(3, 1, 1), motion2, layer2);

	const settings = new Jolt[type]();
	const point = type.startsWith('Swing') ? 'Position' : 'Point';
	settings['m'+point+'1'].Set(offX, 7 + offY, -15+offZ);
	settings['m'+point+'2'].Set(offX, 7 + offY, -15+offZ);

	settings[axis[0]+'1'].Set(0,0,1);
	settings[axis[1]+'1'].Set(1,0,0);

	settings[axis[0]+'2'].Set(1,0,0);
	settings[axis[1]+'2'].Set(0,0,1);

	const constraint = settings.Create(box1, box2);
	physicsSystem.AddConstraint(constraint);
}

let z = -30;
setupBoxes(-20, 0, z, 'FixedConstraintSettings', ['mAxisX','mAxisY'], true, false)
setupBoxes(0, 0, z, 'FixedConstraintSettings', ['mAxisX','mAxisY'], false, true)
setupBoxes(20, 0,z, 'FixedConstraintSettings', ['mAxisX','mAxisY'], false, false)

z+= 16;
setupBoxes(-20, 5, z, 'HingeConstraintSettings', ['mHingeAxis','mNormalAxis',], true, false)
setupBoxes(0, 5, z, 'HingeConstraintSettings', ['mHingeAxis','mNormalAxis'], false, true)
setupBoxes(20, 5, z, 'HingeConstraintSettings', ['mHingeAxis','mNormalAxis'], false, false)

z+= 16;
setupBoxes(-20, 5, z, 'SliderConstraintSettings', ['mSliderAxis','mNormalAxis',], true, false)
setupBoxes(0, 5, z, 'SliderConstraintSettings', ['mSliderAxis','mNormalAxis'], false, true)
setupBoxes(20, 5, z, 'SliderConstraintSettings', ['mSliderAxis','mNormalAxis'], false, false)

z+= 16;
setupBoxes(-20, 5, z, 'SwingTwistConstraintSettings', ['mTwistAxis','mPlaneAxis',], true, false)
setupBoxes(0, 5, z, 'SwingTwistConstraintSettings', ['mTwistAxis','mPlaneAxis'], false, true)
setupBoxes(20, 5, z, 'SwingTwistConstraintSettings', ['mTwistAxis','mPlaneAxis'], false, false)

});
```

**Comment by jrouwe** [comment, marked as answer]
> Is there any practical usage for using non-orthogonal local-space vector combinations? Less than 90 degrees difference per the two vectors? Could that ever do anything useful?

> Is there any practical usage for using non-orthogonal local-space vector combinations? Less than 90 degrees difference per the two vectors? Could that ever do anything useful?

I'm pretty sure that will cause horrible glitches. So no, I don't think it is useful to have the axis at a different angle than 90 degrees.

> If they should always be ortho vectors, I think in most cases I can even just store the two-axis vectors in my files as a Quaternion rotation of stock X/Y axis with predefined assignment, ex: quat applied to (1,0,0) & (0,1,0). Would this make sense? Ex: slider + normal, hinge + normal, twist + plane

I think this is possible, although most people find it a lot easier to work with a couple of vectors than with a quaternion.

> ## Fixed:
> I believe the two provided axis, and thus all 3 resulting axis, are purely to setup initial reference between the two bodies, and there is no real relationship or practical usage to the particular assignment or values? 

Correct.

> Possibly you can just always give global X & Y and I don’t know if any other values would be distinguishable, or always randomly generate any other random two ortho axis? 

Correct.

> If the axis don't match, it looks like it will force a rotation on any non-static object

Correct.

> ## Hinge:
> 2 significant axis given. Hinge axis, and the normal axis: the zero-angle axis, a reference required for defining rotation angles and motor control angle-values

Correct.

> if these axis are different, and one is static, the objects will first orient rotations, then act as expected in the non-fixed axis directions. If they are both dynamic, the object may sometimes orient itself unexpectedly in mid-air.

Correct.

> ## Slider: normal + slider
> I could not find a practical usage for normal axis. Possibly any random orthogonal axis to the slide axis would work and be indistinguishable from any other ortho vector?

It should not really matter what axis you pick indeed, it just needs an orthogonal axis.

> if these axis are different, and one is static, the objects will first orient rotation, then the non-static will support sliding in it's respective axis. If they are both dynamic, the objects will orient in mid-air like fixed constraint.

Correct.

> ## Swing Twist:
> These appear to be structured as orthogonal axis, via plane and twist axis, since internally they calculate a 3rd normal axis. Every provided axis serves a purpose (but still could possibly be defined via a Quat and stock X/Y vectors)

Correct.

> if these axis are different, and one is static, the objects will first orient rotation, then the non-static will support swing-twist in it's respective axis. If they are both dynamic, the objects will orient in mid-air like upside-down fixed constraint.

Correct.

**Reply by jrouwe** [reply]
B.t.w. the idea of storing the vectors as a quaternion would work if you specify constraints in world space. If you specify them in local space then you would need 2 quaternions.

**Reply by PhoenixIllusion** [reply]
I'd intended the quaternion to respect local-space / world-space of the connected body using the constraints. In the case of the local-space setup, I'd configure the X/Y axis first to local-body-space before applying the Quat. I believe that is the 2 quaternions used that you are referencing?

The usage of the quaternions is that wiring up an editor to two distinctly controlled axis vectors is a bit more challenging than wiring up a a static 2-axis indicator and connected Quat and then converting it prior to usage into the guaranteed orthogonal vectors.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/eb6d81a6-2948-4e86-a0ea-94131f007f90)

**Reply by PhoenixIllusion** [reply]
> I think this is possible, although most people find it a lot easier to work with a couple of vectors than with a quaternion.

This was purely for "I was thinking of solving my own personal problems this way with my GUI editor", not anything about the existing approach

**Reply by jrouwe** [reply]
> I believe that is the 2 quaternions used that you are referencing?

No what I meant is that if you don't know the world positions of the bodies yet then you need 2 quaternions: 1 to represent the rotation to go from body 1 space to constraint 1 space and one to go from body 2 space to constraint 2 space. If the bodies are already in world space then you only need 1 quaternion as the relative orientation between the bodies is already known.

> I was thinking of solving my own personal problems this way with my GUI editor

Yes, I understood that :)
